### ISSUE TITLE
Unsupported flag action ROL

### ISSUE THREAD

**Issue by extremecoders-re** [angr/angr#69]
I was testing the following assembly code compiled with fasm.

```nasm
format PE

entry start

section '.text' code readable executable

start:
	mov ecx, 10

next:
	rol eax, 1
	loop next

check:
	cmp eax, 0xDEADC0DE
	jz success

failure:
	xor eax, eax
	ret

success:
	mov eax, 1
	ret
```
Basically, I want to find the initial value of `eax` such that after 10 `rol eax, 1` operations its value would be `0xDEADC0DE`. (The answer is `0x37B7AB70`)

The disassembly of the code looks like this:
![bb](https://cloud.githubusercontent.com/assets/12072382/13956553/40afe3d0-f06f-11e5-9bfe-bc5335a884de.png)

and the script is as follows.

```python
#!/usr/bin/python

import angr

def main():
    proj = angr.Project('test.exe')

    initial_state = proj.factory.blank_state(addr=0x401000)
    r_eax = initial_state.se.BVS('eax', 32)
    initial_state.regs.eax = r_eax

    pg = proj.factory.path_group(initial_state, immutable=False)

    pg.explore(find=0x401013, avoid=0x401010)
    found_state = pg.found[0].state

    print found_state.se.any_int(r_eax)

if __name__ == '__main__':
    main()
```

However on running I am getting the following error

```
WARNING | 2016-03-22 20:38:46,984 | cle.pe | The PE module is not well-supported. Good luck!
ERROR   | 2016-03-22 20:38:47,076 | simuvex.vex.ccall | Unsupported flag action ROL
```

Is the `rol` instruction unsupported ? 

(The binary is provided for reference: https://drive.google.com/open?id=0B4nawd5TCX1cakhaLTROMW5XdXM)

**Comment by ltfish** [comment]
@zardus

**Comment by extremecoders-re** [comment]
**Additional information:**
I think the problem has some relation with the way the `loop` instruction is handled. For example if i replace the `loop` instruction with the following 3 instruction sequence; angr works just fine.

```nasm
format PE

entry start

section '.text' code readable executable

start:
	mov ecx, 10

next:
	rol eax, 1
	dec ecx             ; <<<<<<<<<<<<<<<<<<<<<<<<<<
	test ecx, ecx       ; <<<<<<<<<<<<<<<<<<<<<<<<<<
	jnz next            ; <<<<<<<<<<<<<<<<<<<<<<<<<<

check:
	cmp eax, 0xDEADC0DE
	jz success

failure:
	xor eax, eax
	ret

success:
	mov eax, 1
	ret
```

**Comment by ltfish** [comment]
The problem is that having a conditional jump following an `ROL` operation is not supported by angr right now. It shouldn't be difficult to add the support, we just need someone to write the code (~10 lines) in `simuvex/s_ccall.py`. That's why I @zardus before, since he has a much better idea of those code than me.

**Comment by ltfish** [comment]
And thanks for properly formatting your issue - it looks really nice!

**Comment by zardus** [comment]
I actually have an experimental ROL action implementation locally -- I'll dust it off later today and test it with your sample.

**Comment by extremecoders-re** [comment]
@zardus 

Thanks.
So the problem is that the effect of the `ROL` instruction on the the `EFLAGS` is not implemented. Hence if an immediate instruction uses the `EFLAGS`, it triggers the error.

**Comment by ltfish** [comment]
@extremecoders-re that's right.

**Comment by extremecoders-re** [comment]
I Implemented the missing [pc_actions_ROL](https://github.com/angr/simuvex/blob/master/simuvex/vex/ccall.py#L375) function in simuvex/ccall.py.

```python
def pc_actions_ROL(state, nbits, res, _, cc_ndep, platform=None):
    cf = res[0]
    pf = (cc_ndep & data[platform]['CondBitMasks']['G_CC_MASK_P'])[data[platform]['CondBitOffsets']['G_CC_SHIFT_P']]
    af = (cc_ndep & data[platform]['CondBitMasks']['G_CC_MASK_A'])[data[platform]['CondBitOffsets']['G_CC_SHIFT_A']]
    zf = (cc_ndep & data[platform]['CondBitMasks']['G_CC_MASK_Z'])[data[platform]['CondBitOffsets']['G_CC_SHIFT_Z']]
    sf = (cc_ndep & data[platform]['CondBitMasks']['G_CC_MASK_S'])[data[platform]['CondBitOffsets']['G_CC_SHIFT_S']]
    of = ((res >> (nbits - 1)) ^ res)[0]
    return pc_make_rdata(data[platform]['size'], cf, pf, af, zf, sf, of, platform=platform)
```

The `ROL` instruction only affects the `CF` & `ZF`. 
```
CF = lsb(result)
OF = msb(result) ^ lsb(result)
```
Remaining flags are unaffected. With the above patch there aren't any further errors and the script works.

Let me know, I if would send a PR.

**Comment by ltfish** [comment]
@extremecoders-re thanks! It seems reasonable to me (if your ROL description is correct).

@zardus your code won't be used :-)

**Comment by zardus** [comment]
Wow, thanks! Sorry I didn't get to it; we've been struggling with yet another deadline.

A PR would be amazing. If it's not too much trouble, could you also add a PR to the `binaries` repo with your example binary that triggers this issue, and add a `test_rol.py` to the `angr` repo that makes sure all of this works? If that's too much trouble, just the `simuvex` PR will work, but then the code will probably break later due to the lack of a testcase :-)

One comment on the code itself: `>>` is an arithmetic right shift. It shouldn't matter here, as I don't think `nbits` can ever be greater than the size of `res`, but it might be good to use `state.se.LShR(res, nbits-1)` (logical shift right) rather than the arithmetic shift.

**Comment by extremecoders-re** [comment]
@zardus Yes, it should be `LShr` to be on the safer side, I will change that and create a PR with everything you mentioned.

**Comment by zardus** [comment]
Thanks! Sorry to load you with all these PRs; no good deed goes unpunished ;-)

**Comment by ronnychevalier** [comment]
There is already `test_vex.py` with the function `test_ccall`, you
should add your test in here instead of a new file.

**Comment by zardus** [comment]
I'd like to have a real-life binary test for it, as we seem to be in the habit of modifying VEX for our purposes, but adding it to `test_ccall` is a good idea as well.

**Comment by extremecoders-re** [comment]
Closing this issue as it has been fixed with simuvex commit [e56ba15](https://github.com/angr/simuvex/commit/e56ba15594c0b9c3cda480aa2f137bdbe29b31f6)
